kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/aggregate/EventSourcingAggregateWithImmutableIdentifierRepository.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2020. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.axonframework.extensions.kotlin.aggregate
+
+import org.axonframework.eventsourcing.EventSourcedAggregate
+import org.axonframework.eventsourcing.EventSourcingRepository
+import org.axonframework.messaging.unitofwork.CurrentUnitOfWork
+import org.axonframework.modelling.command.Aggregate
+import org.axonframework.modelling.command.LockAwareAggregate
+import java.util.concurrent.Callable
+
+/**
+ * Event souring repository which uses a aggregate factory to create new aggregate passing <code>null</code> as first event.
+ * @param builder repository builder with configuration.
+ *
+ * @since 0.2.0
+ * @author Simon Zambrovski
+ */
+class EventSourcingAggregateWithImmutableIdentifierRepository<A>(
+    builder: Builder<A>
+) : EventSourcingRepository<A>(builder) {
+
+    override fun loadOrCreate(aggregateIdentifier: String, factoryMethod: Callable<A>): Aggregate<A> {
+        val factory = super.getAggregateFactory()
+        val uow = CurrentUnitOfWork.get()
+        val aggregates: MutableMap<String, LockAwareAggregate<A, EventSourcedAggregate<A>>> = managedAggregates(uow)
+        val aggregate = aggregates.computeIfAbsent(aggregateIdentifier) { aggregateId: String ->
+            try {
+                return@computeIfAbsent doLoadOrCreate(aggregateId) {
+                    // call the factory and instead of newInstance on the aggregate class
+                    factory.createAggregateRoot(aggregateId, null)
```

Comment:
I've seen users provide their own `AggregateFactory` to set the identifier directly without changing anything on the `EventSourcingRepository`. For the time being I fail to follow why this is necessary. Would you mind explaining thisin more detail then just the comment above?

Reply:
As far as I could understand the code, the method on `AbstractRepository#loadOrCreate` is called. In the end, the problem of the default implementation is located inside `AggregateAnnotationCommandHandler#initializeHandler()`. This private method decides based on the `AggregateCreationPolicy` and sets the factory method for creation of new aggregates to default parameterless constructor:

``` 
                    case CREATE_IF_MISSING:
                        handlersFound.add(new AggregateCreateOrUpdateCommandHandler(
                                handler, aggregateModel.entityClass()::newInstance
                        ));
                        break;
``` 
The easiest way to change this behaviour that I found was to overwrite the Repository...

Further more, I believe it is a mistake that the command handler decides in a private method about how the new instance is created (having repository and factory abstractions I expect to delegate to one of those classes).

Reply:
The easiest way to find this, is to replace the implementation by the call of `super` in my repository and setup the break point into the initializeHandler method.